### ISSUE TITLE
Unusual extreme memory usage

### ISSUE THREAD

**Issue by HipyCas** [navidrome/navidrome#2003]
### Description

I was checking my server system stats and was surprised by a really high RAM usage. So, I went to check what was going on, and saw a whooping 11.8GB  usage by Navidrome. I went to check the Docker logs. It was not  scanning (as far as I believe) and nobody was using it.

### Expected Behaviour

Just consume the RAM that it always consume, restarting the container actually fixes the problem.

### Steps to reproduce

Would love to be able to provide details, but it is a by the book installation. Currently its been up 47 days and has 429 songs.

### Platform information

 - Navidrome version: _hehe, just updated so idk_
 - Browser and version: _Not applicable_
 - Operating System: Debian 11.5 with Docker 20.10.19 and Compose 2.11.2

### Additional information

The data from `docker stats`:
![2022-11-20_20-09](https://user-images.githubusercontent.com/54776315/202921765-df5dd75d-c33e-47f5-b34b-3d7f8c924f4d.png)

**Comment by deluan** [comment]
So it was with a previous version, correct? Let me know if you run into this issue with the latest (0.48.0) version.

**Comment by HipyCas** [comment]
I'll keep an eye on the memory usage and come back to this if the is a memory spike again.

**Comment by deluan** [comment]
Could this be a duplicate of https://github.com/navidrome/navidrome/issues/1874 ?

**Comment by strider-** [comment]
> Could this be a duplicate of #1874 ?

I'll give the latest dev build a shot (`develop` tag in Docker?) and see if it addresses the issue

**Comment by HipyCas** [comment]
> Could this be a duplicate of https://github.com/navidrome/navidrome/issues/1874 ?

I don't think so, the high memory usage was happening for a long time and I don't remember scanning the library while then.

**Comment by deluan** [comment]
> I don't think so, the high memory usage was happening for a long time and I don't remember scanning the library while then.

The default is to scan every minute, unless you changed the `ScanSchedule` config option.

**Comment by strider-** [comment]
After running the dev build for a few days (Version: 0.48.0-SNAPSHOT (db6e8e45)) it looks like the issue is still persisting :(

![image](https://user-images.githubusercontent.com/651313/210101594-fcf97316-c081-413a-9de9-e64e2adabaf6.png)

**Comment by HipyCas** [comment]
It's happening again, not using it rn and neither UI nor docker logs show anything about scanning the library. Anywhere that I can look to get more info and maybe know where the ram usage comes from?

![image](https://user-images.githubusercontent.com/54776315/212492086-70543bfe-815e-47a2-bec2-289f22abd3a2.png)

**Comment by dessalines** [comment]
I'm also getting this... very strange spikes in CPU and memory after navidrome has been running for a few days.

**Comment by xr09** [comment]
I'm experiencing something similar. Even though my Navidrome instance is mostly idle (I'm the only user) and the library is very small, I'm surprised to see the memory usage has been growing at a steady pace for the last two months. 

![Screenshot_20230123_194323](https://user-images.githubusercontent.com/1380243/214124206-c35b6ae0-de01-45eb-b8e1-9653b27af12c.png)

**Comment by JulienSambre** [comment]
Hi!

I installed the docker version (deluan/navidrome:0.48.0) of Navidrome a few days ago (moving from Funkwhale) and so far it works like a charm (really a good piece of software comparing at the alternatives), except this big memory leak.

![image](https://user-images.githubusercontent.com/14327288/215420797-0ab68325-d53f-46b5-a545-6d783acd0621.png)

I set nothing in the container except the log level, so Navidrome scan my library (almost 30 Go) every 1 minute.

I noticed several things :
* At first I didn't fixed the max available memory of the container in order to see how much it could needed. The memory grew like everybody
* Then, I fixed the max available memory of the container to 200 Mo RAM => it seemed to work when Navidrome almost reached the limit (it released like 40 Mo) but then it started again to grow
* When the memory limit was reached, the container didn't crashed, but it was unavailable (unable to reached the UI). In the same time, I saw a strong increase of my IO waits (my music is bridged on the VM via a NAS) : it seemed to be caused by the music scan, I don't know why

For now, I will put the max memory to 300 Mo (should be enough for me) and restart the container every day. Not wonderful, a fix would be fantastic, seems to be a very classic memory leak issue ;).

**Comment by cceckman** [comment]
A little bit of data:

I've also seen slow-but-steady increases in memory usage while Navidrome is running on my system (Raspberry Pi 3, NixOS). I'm on the default schedule of `@every 1m` for scanning.

![Total memory usage increasing, drops at restarts](https://user-images.githubusercontent.com/861696/219953580-f3432973-a982-462a-82ae-da91001bc64c.png)
![AnonPages memory increasing](https://user-images.githubusercontent.com/861696/219954368-ae9e8b40-e8ad-4747-98b5-5565a8ae9a7c.png)

I tweaked my Navidrome instance to include the [pprof http handler](https://pkg.go.dev/net/http/pprof#Index) ([commit](https://github.com/cceckman/navidrome/commit/f7499096f2b8cf6ba4269ebfaf37954f104e0a1e)). (That's not pull-ready, just a hack.) I took an `allocs` sample a few minutes after starting the process, and another this morning, to try to see where the memory is going.

The `pprof` snapshots are, so far as I can tell, the same. I'll check again after having run longer.

**Comment by cceckman** [comment]
I realize there was a confounding factor on my system: I had Prometheus running on the same node, which was eating at least some memory / pages / etc. I've restarted everything at Navidrome 0.48.0, and am keeping a log of `system.slice/navidrome.service`'s memory usage (asking `systemd-cgtop`) as well as node-wide usage (from `node_exporter`).

**Comment by cceckman** [comment]
A _little_ more specific data: here are the outputs from some hourly crons, invoking `systemd-cgtop` on Navidrome's slice in particular, and on the system as a whole. This is with "normal" usage - there are at most two users, and we're using it fairly lightly.

[the Navidrome slice in particular](https://github.com/navidrome/navidrome/files/10947629/navidrome-container-monitor.log), and (for part of the time) on [the system as a whole](https://github.com/navidrome/navidrome/files/10947630/containers.log). Navidrome's usage is steadily, though not monotonically, increasing. No other slice seems to be occupying much.

This isn't a "patched with pprof handler" version, it's just Navidrome 0.48 as build under Nix.

I can say: It doesn't appear the DB is contributing significantly; at latest read:

```
rack3% sudo ls -lah /var/lib/private/navidrome
total 41M
drwxr-xr-x 2 navidrome navidrome 4.0K Feb 20 16:18 .
drwx------ 3 root      root      4.0K May 21  2022 ..
-rw------- 1 navidrome navidrome  37M Mar 10 23:20 navidrome.db
-rw------- 1 navidrome navidrome  32K Mar 10 23:49 navidrome.db-shm
-rw------- 1 navidrome navidrome 4.2M Mar 10 23:49 navidrome.db-wal
```

So- it sure seems like memory is going _somewhere_ in the Navidrome container.

**Comment by deluan** [comment]
Hey @cceckman , thanks for the information. Can you try the same, but with the scanner disabled (`ScanSchedule = "0"`)?

**Comment by cceckman** [comment]
Sure thing. Next batch of changes under test:

- [Navidrome 0.49.3](https://github.com/NixOS/nixpkgs/commit/ee89f8e1e558c2021a7bea8a7dd50b9521afb772)
- [Disabled automated scan](https://github.com/cceckman/homelab/commit/d88113dd75a1e099811784288363dc2580c0bcb9)
- Scrape via HTTPS - trying to get my Prometheus instance to hit the `/metrics` endpoint properly, to get that data stream as well.

Will report back after it's run a while.

**Comment by cceckman** [comment]
With the above - much more stable: [navidrome-container-monitor.log](https://github.com/navidrome/navidrome/files/11000916/navidrome-container-monitor.log)

~150 MiB idle after startup (and maybe after an initial scan?), with no usage; I see that doubled to ~338 MiB when I played some music and stayed there, but isn't "just" increasing.

The Prometheus metrics don't show anything super interesting over time. `go_memstats` and `process_resident_memory_bytes` both show much smaller values than systemd (`process_resident_memory_bytes` < 60MiB)- but if I'm reading right, the cgroup numbers also covering some portion of file / cache that `/metrics` don't cover. So that's not really a concerning difference.

I have a [tweak](https://github.com/cceckman/homelab/commit/c557002102c1d2237afa0964eefb64bf981921c7) to the cron-monitor to capture the detailed meminfo (`/sys/fs/cgroup/system.slice/navidrome.service/memory.stat`).

Should I reenable the automated scan to see where the ballooning is accounted to? Or trigger some manual scans to see what happens?

**Comment by cceckman** [comment]
Back with more data- the CSV file with samples of `memory.stat` from above: [navidrome-container-memstat.csv](https://github.com/navidrome/navidrome/files/11035238/navidrome-container-memstat.csv) First column is Unix Epoch, the rest of the fields are from sysfs.

It seems pretty stable for a couple days... until, due to New Music (yay!) we ran a scan. 

---

We ran a manually-triggered quick scan on 2023-03-18 ~10:43, and then a full scan at ~10:55 (US Eastern Time.) After that, we've had 2-3 streaming clients via web UI or DSub.

Eyeballing a chart from that CSV - the "file" counters come up during the scan, and never go _all_ the way down. In particular, the `active_file` count never entirely goes back down.

Two other data sets, from the Prometheus metrics (note: Eastern Time timestamps, so offset from epoch time):

![Chart of Open Descriptors](https://user-images.githubusercontent.com/861696/226775265-a0d8ae49-8aa6-4db2-b458-c4ab8a7eafb9.png)

![Chart of Threads](https://user-images.githubusercontent.com/861696/226775846-047aa474-42c6-4ab4-b3b8-e94b7642190a.png)

The "raise-to-plateau" of both of those align with the triggered scan, ~10:43.

**Comment by johndoe0815** [comment]
I can add on this.
I have Navidrome running in a Proxmox container, mostly idle, not using it much with only me as only user. I gave the container 1GB of RAM, and this was slowly eaten up over time:
![grafik](https://github.com/navidrome/navidrome/assets/92877294/0733dce1-d1f3-487b-bfe8-01bddc6cc14d)
I then restarted the container, reducing the memory consumption drastically, but without having accessed Navidrome even once since then, it is again slowly increasing.
There must be any memory issue which at the moment would require to restart the container from time to time.
I do not have this with other services running in the same kind of way.

**Comment by johndoe0815** [comment]
Is there any solution in sight?
I keep restarting the container from time to time, but even without using Navidrome at all, memory consumption constantly increases over time until next restart:
![grafik](https://github.com/navidrome/navidrome/assets/92877294/06b6af82-e458-4ba2-8f2f-08bf98777596)

**Comment by markasoftware** [comment]
I don't believe any solution is in sight. I'm experiencing exactly the same thing as you @johndoe0815 I've resorted to using systemd to automatically restart the process when it uses too much memory. https://xkcd.com/1495/

**Comment by mybuntu-git2** [comment]
I had the same issue using docker on ARM, used the same workaround (scan schedule disabled, scan via CLI, restart navidrome periodically) and the issue seems fixed in 0.50.1

**Comment by deluan** [comment]
Thanks @mybuntu-git2 for reporting back. I did some investigation a while ago and I think this is not really an issue with Navidrome, but how Go's Garbage Collector (GC) works. More info here:

 - https://weaviate.io/blog/gomemlimit-a-game-changer-for-high-memory-applications
 - https://github.com/golang/go/issues/14521#issuecomment-189244596
 
Based on the info above, doesn't seem it is a memory leak, it is just the GC being "lazy" and leaving the work for when memory is actually needed. The information from the posts above indicates that the solution would be to limit the maximum amount of memory the Navidrome process "sees", so the GC would start reclaiming memory earlier and never going past the specified limit. I did some experimentation, but unfortunately I couldn't get to any conclusive results.

Now seems that in Go 1.21 (used in 0.50.1), some changes in the GC behaviour [were introduced](https://tip.golang.org/doc/go1.21#runtime-changes) that could change how it affects long running applications like Navidrome.

If anyone experiencing this could validate the "memory leak" is gone, please let me know, so we can close this.

**Comment by mintsoft** [comment]
I'm using 0.50.1 and have the same sort of behaviour as a few others here. 

Limited use instance, running in a container (LXC rather than docker), single user; RAM climes steadily and consistently up even when not being actively used. At approx 1.1GB I get alerts for the container and then restart the process and the cycle begins again.

**Comment by mintsoft** [comment]
I've just had to restart navidrome because it had gobbled up most of the RAM on the host again; I'm just testing it with the added Environmental variable `GOMEMLIMIT=300000000`; which _should_ validate the comment here: https://github.com/navidrome/navidrome/issues/2003#issuecomment-1858885696

If basically it starts behaving better, then it's likely it's golang not returning the memory to the OS.

**Comment by johndoe0815** [comment]
@deluan 
Unfortunately no improvement also with version 50.2. Last restart 21.12., no active usage from of Navidrome from an end users since then:
![grafik](https://github.com/navidrome/navidrome/assets/92877294/9263b019-323f-48c2-b801-040c293d45a1)

**Comment by markasoftware** [comment]
I've also been using Navidrome 0.50.2 since it was released and the memory issues continue. I run it in a cgroup with memory limited to 250 MiB, so this should definitely trigger Go's garbage collector. However it is not good enough, and my process OOMs and then restarts at least once per week.

**Comment by f-bn** [comment]
From my knowledge, since memory limiting and reporting with Docker/LXC (or whatever container tech on Linux) is done using control groups, the memory usage is accounting cache usage of the container as part of memory used depending on the cgroup stat file, therefore not all the program using cgroup renders it the same way.

```shell
$ egrep "rss|cache" /sys/fs/cgroup/memory/docker/8de978f32abe0384869ef270cbb7e349bc40d7c430cd22f753c99411829e5090/memory.stat
cache 433336320 # 433MiB
rss 26042368 # 26MiB
rss_huge 0
total_cache 433336320
total_rss 26042368
total_rss_huge 0

$ cat /sys/fs/cgroup/memory/docker/8de978f32abe0384869ef270cbb7e349bc40d7c430cd22f753c99411829e5090/memory.usage_in_bytes
461443072 # 461 MiB

$ docker stats -a
...
8de978f32abe   navidrome      0.00%     163.7MiB / 4GiB       4.00%     0B / 0B   0B / 0B     0
```

So I don't think this is the ideal metrics to follow what's going on regarding memory usage. The cache usage is obviously important since the program is reading files most of the time.

**Comment by mintsoft** [comment]
Just to clarify the situation for my test case vs the previous comment. Although I am putting navidrome within LXC, I am not using cgroups or applying any memory limit to the container itself. The actual host only has 4GB of RAM. So this is a  more "real" test of the memory behaviour imo

**Comment by mintsoft** [comment]
Just started getting notifications again for 75% RAM on the host with GOMEMLIMIT set :(
```
● navidrome.service - Navidrome Music Server and Streamer compatible with Subsonic/Airsonic
     Loaded: loaded (/etc/systemd/system/navidrome.service; enabled; preset: enabled)
    Drop-In: /etc/systemd/system/navidrome.service.d
             └─override.conf
     Active: active (running) since Thu 2024-01-11 09:33:27 UTC; 1 week 3 days ago
   Main PID: 33592 (navidrome)
      Tasks: 12 (limit: 2302)
     Memory: 699.5M
        CPU: 3d 16h 52min 58.407s
     CGroup: /system.slice/navidrome.service
             └─33592 /opt/navidrome/navidrome --configfile /var/lib/navidrome/navidrome.toml

```
Took  10 days, to hit that limit.

htop:
![Screenshot_20240121_104033](https://github.com/navidrome/navidrome/assets/978048/ce05ddac-5600-4717-973e-a0d052b47225)


And that was with this:
```
root@navidrome:/# systemctl show navidrome | grep Env
Environment=GOMEMLIMIT=300000000
```

**Comment by mybuntu-git2** [comment]
Reporting more details about my case, hoping this helps: 
Ubuntu 22.04 ARM (Odroid N2+)
navidrome version 0.50.2

I run navidrome in docker, no alterations to gomemlimit.

with the following setup options relevant for memory usage (and Spotify and Lastfm set up, not included below)

```
>      ND_SCANSCHEDULE: "@every 60m"
>       ND_LOGLEVEL: error
>       ND_SESSIONTIMEOUT: 30m
>       ND_COVERARTPRIORITY: "external, embedded, cover.*, folder.*, front.*"
>       ND_ENABLECOVERANIMATION: "false"
>       ND_ENABLEMEDIAFILECOVERART: "false"
>       ND_IMAGECACHESIZE: "0"
>       ND_JUKEBOX_ENABLED: "false"
>       ND_LISTENBRAINZ_ENABLED: "false"
>       ND_SCANNER_GROUPALBUMRELEASES: "true"
>       ND_SCANNER_EXTRACTOR: "ffmpeg"
```

Navidrome uses 23MB idle according to docker stats.
Memory spikes to max 200MB when in use,   and then is deallocated successfully.
my system flushes caches in cron every few hours.

I have around 2K songs and two users.

**Comment by mintsoft** [comment]
Something I've just noticed (after restarting navidrome) ps vs systemd's understanding of the memory is different:
```
root@navidrome:/# systemctl status navidrome
● navidrome.service - Navidrome Music Server and Streamer compatible with Subsonic/Airsonic
     Loaded: loaded (/etc/systemd/system/navidrome.service; enabled; preset: enabled)
    Drop-In: /etc/systemd/system/navidrome.service.d
             └─override.conf
     Active: active (running) since Sun 2024-03-17 10:23:46 UTC; 4min 19s ago
   Main PID: 62333 (navidrome)
      Tasks: 12 (limit: 4640)
     Memory: 241.3M
        CPU: 36.030s
     CGroup: /system.slice/navidrome.service
             └─62333 /opt/navidrome/navidrome --configfile /var/lib/navidrome/navidrome.toml

```

vs
```
root@navidrome:/# ps aux | grep navidrome
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
navidro+   62333 12.1  1.2 2086912 50012 ?       Ssl  10:23   0:43 /opt/navidrome/navidrome --configfile /var/lib/navidrome/navidrome.toml
```

**Comment by mintsoft** [comment]
I've just built pprof into my local navidrome and I'll run it for a few weeks to see if this sheds any sort of light on the behaviour. If anyone else wants to do this, the easiest way I've found to enable pprof is through: https://github.com/mintsoft/navidrome/commit/04e6cd21c7a11c7a23647777b402616faf6ba7a7

**Comment by deluan** [comment]
> I've just built pprof into my local navidrome and I'll run it for a few weeks to see if this sheds any sort of light on the behaviour. If anyone else wants to do this, the easiest way I've found to enable pprof is through: [mintsoft@04e6cd2](https://github.com/mintsoft/navidrome/commit/04e6cd21c7a11c7a23647777b402616faf6ba7a7)

Actually, the easiest way is to use the undocumented config option `DevEnableProfiler=true`: https://github.com/navidrome/navidrome/blob/7c5eec715ddeab36f3b5dc4a0d4db5beebffe9a0/cmd/root.go#L115-L117

I put it in place when I was investigating this issue, but couldn't find anything relevant. It is good to have more eyes on this, though. Thanks for taking a stab on it.

**Comment by BaeKey** [comment]
> Thanks @mybuntu-git2 for reporting back. I did some investigation a while ago and I think this is not really an issue with Navidrome, but how Go's Garbage Collector (GC) works. More info here:
> 
> * https://weaviate.io/blog/gomemlimit-a-game-changer-for-high-memory-applications
> * [runtime: GC: heap idle is not released / decrease golang/go#14521 (comment)](https://github.com/golang/go/issues/14521#issuecomment-189244596)
> 
> Based on the info above, doesn't seem it is a memory leak, it is just the GC being "lazy" and leaving the work for when memory is actually needed. The information from the posts above indicates that the solution would be to limit the maximum amount of memory the Navidrome process "sees", so the GC would start reclaiming memory earlier and never going past the specified limit. I did some experimentation, but unfortunately I couldn't get to any conclusive results.
> 
> Now seems that in Go 1.21 (used in 0.50.1), some changes in the GC behaviour [were introduced](https://tip.golang.org/doc/go1.21#runtime-changes) that could change how it affects long running applications like Navidrome.
> 
> If anyone experiencing this could validate the "memory leak" is gone, please let me know, so we can close this.

My Docker Compose configuration file is as follows

```
services:
  navidrome:
    image: deluan/navidrome:latest
    ports:
      - "4533:4533"
    restart: unless-stopped
    environment:
      ND_SCANSCHEDULE: 24h
      ND_LOGLEVEL: error
      ND_SESSIONTIMEOUT: 24h
      ND_ENABLEEXTERNALSERVICES: false
      ND_ENABLEDOWNLOADS: false
    volumes:
      - "./data:/data"
      - "./music:/music:ro"
    deploy:
      resources:
        limits:
          memory: 64M
```

By setting

```
    deploy:
      resources:
        limits:
          memory: 64M
```

and then using 'docker stats' to check memory usage, I found that Go's Garbage Collector (GC) works, with memory usage stabilizing at around 30M.
